src/man/man8/sefreport.8
.\" $Revision: 1.22 $
.ds ]W Sun Microsystems
.\" SAM-QFS_notice_begin
.\"
.\" CDDL HEADER START
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
.\" or https://illumos.org/license/CDDL.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at pkg/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\" CDDL HEADER END
.\"
.\" Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
.\" Use is subject to license terms.
.\"
.\" SAM-QFS_notice_end
.nh
.na
.TH sefreport 8 "3 Apr 2001"
.SH NAME
sefreport \- Displays the content of the System Error Facility (SEF) log
.SH SYNOPSIS
\fB/opt/SUNWsamfs/sbin/sefreport\fR
\%[\fB\-v|-t\fR]
\%\fB\-d\fR
\%\fIfile\fR
.SH AVAILABILITY
\fBSUNWsamfs\fR
.SH DESCRIPTION
The \fBsefreport\fR command 
reads the content of a Sun StorEdge \%SAM-FS or Sun \%SAM-QFS SEF log file
and writes its output
to \fBstdout\fR in a \%human-readable format.
By default, the log file
is \fB/var/opt/SUNWsamfs/sef/sefdata\fR.
The SEF log file contains the data gathered from the log
sense pages of peripheral tape devices used by Sun StorEdge \%SAM-FS
and Sun \%SAM-QFS file systems. 
For more information on the SEF log file, including
its content and format, see the \fBsefdata\fR(5) man page.
.PP
The \fBsefreport\fR command reads the input file specified by
the \fIfile\fR argument.
If no other options are specified,
the \fBsefreport\fR command examines the SEF log file and generates
the following information for each record contained in \fIfile\fR:
.TP 2
\(bu
The first header line states the record number,
which is its ordinal position in the file.
.TP
\(bu
The second header line contains the timestamp of the record,
the vendor name of the device from
which the log sense data was received, the product name of the device,
the revision level of the device's firmware, the
string \fBVSN\fR, and the Volume Serial Name (VSN) of the volume
mounted in the device
when the log sense data was generated.
.PP
Following the header lines, the log sense data for each page in the record
is printed.  For each log sense page, a line identifying the page code
is printed, followed by a line of column headings.  The data is then
printed in three columns per line with the following headings:
parameter code, control, and
parameter value.  All data is generated in hexadecimal notation.
For the meanings of
the parameter codes, control bits, and parameter values, see your
vendor
documentation for the specific device.
.SH OPTIONS
This command accepts the following options:
.TP 5
\fB\-d\fR
Includes additional device information.
For each record, the command generates a third header line
that identifies the equipment number
of the device as configured in the \fBmcf\fR file and the path name of the
device.
.TP
\fB\-t\fR
Generates log sense output with text descriptions.
On each line of log sense data output, an additional string containing
the equipment number, page code, VSN, and parameter code description is
printed.  The -t option is not used when the -v option is specified. 
.TP
\fB\-v\fR
Generates verbose output.
On each line of log sense data output, an additional string containing
the equipment number, page code, and VSN is printed.  This string is
enclosed in parentheses and the items are \%colon-separated.
.SH OPERANDS
This command accepts the following operand, which must be
specified:
.TP 10
\%\fIfile\fR
Specifies the SEF log file.
The SEF log file can be read from its default
location (\fB/var/opt/SUNWsamfs/sef/sefdata\fR)
or it can be redirected to another file for SEF processing.
.SH EXAMPLES
Example 1.
Assume that your system is set up to write SEF values
to file \fB/var/opt/SUNWsamfs/sef/sefdata.mid\fR.
You have entered the following
command to write the SEF data using the report formatter:
.PP
.ft CO
.nf
srvr# sefreport /var/opt/SUNWsamfs/sef/sefdata.mid > ~mydir/sef.short
.fi
.ft
.PP
The file \fB~mydir/sef.short\fR is as follows:
.PP
.ft CO
.nf
Record no. 1
Mon Mar 26 11:17:48 2001  STK      9840             1.25 VSN 002981

   PAGE CODE 2
   param code  control   param value
       00h       74h     0x0
       01h       74h     0x0
       02h       74h     0x0
       03h       74h     0x0
       04h       74h     0x0
       05h       74h     0x40050
       06h       74h     0x0


   PAGE CODE 3
   param code  control   param value
       00h       74h     0x0
       01h       74h     0x0
       02h       74h     0x0
       03h       74h     0x0
       04h       74h     0x0
       05h       74h     0x140
       06h       74h     0x0


   PAGE CODE 6
   param code  control   param value
       00h       74h     0x0


Record no. 2
Mon Mar 26 11:30:06 2001  STK      9840             1.25 VSN 002999

   PAGE CODE 2
   param code  control   param value
       00h       74h     0x0
       01h       74h     0x0
       02h       74h     0x0
       03h       74h     0x0
       04h       74h     0x0
       05h       74h     0x1400a0
       06h       74h     0x0


   PAGE CODE 3
   param code  control   param value
       00h       74h     0x0
       01h       74h     0x0
       02h       74h     0x0
       03h       74h     0x0
       04h       74h     0x0
       05h       74h     0x190
       06h       74h     0x0


   PAGE CODE 6
   param code  control   param value
       00h       74h     0x0

<<<NOTE:  This output has been truncated for inclusion on this
man page.>>>
.fi
.ft
.PP
Example 2:
Assume that you also need to produce a report with additional data.
You can use the same log file as in Example 1, but
you want this report to contain more
information than \fBsef.short\fR, so you
invoke \fBsefreport\fR with the \fB\-d\fR and \fB\-v\fR options.
The following
command is entered:
.PP
.ft CO
.nf
srvr# sefreport -d -v /var/opt/SUNWsamfs/sef/sefdata.mid > ~mydir/sef.long
.fi
.ft
.PP
The file \fB~mydir/sef.long\fR is as follows:
.PP
.ft CO
.nf

Record no. 1
Mon Mar 26 11:17:48 2001  STK      9840             1.25 VSN 002981
   Eq no. 32   Dev name /dev/rmt/1cbn

 rec  pg cd   param code  control   param value
   1    2         00h       74h     0x0          (32:2:002981)
   1    2         01h       74h     0x0          (32:2:002981)
   1    2         02h       74h     0x0          (32:2:002981)
   1    2         03h       74h     0x0          (32:2:002981)
   1    2         04h       74h     0x0          (32:2:002981)
   1    2         05h       74h     0x40050      (32:2:002981)
   1    2         06h       74h     0x0          (32:2:002981)


 rec  pg cd   param code  control   param value
   1    3         00h       74h     0x0          (32:3:002981)
   1    3         01h       74h     0x0          (32:3:002981)
   1    3         02h       74h     0x0          (32:3:002981)
   1    3         03h       74h     0x0          (32:3:002981)
   1    3         04h       74h     0x0          (32:3:002981)
   1    3         05h       74h     0x140        (32:3:002981)
   1    3         06h       74h     0x0          (32:3:002981)


 rec  pg cd   param code  control   param value
   1    6         00h       74h     0x0          (32:6:002981)


Record no. 2
Mon Mar 26 11:30:06 2001  STK      9840             1.25 VSN 002999
   Eq no. 31   Dev name /dev/rmt/0cbn

 rec  pg cd   param code  control   param value
   2    2         00h       74h     0x0          (31:2:002999)
   2    2         01h       74h     0x0          (31:2:002999)
   2    2         02h       74h     0x0          (31:2:002999)
   2    2         03h       74h     0x0          (31:2:002999)
   2    2         04h       74h     0x0          (31:2:002999)
   2    2         05h       74h     0x1400a0     (31:2:002999)
   2    2         06h       74h     0x0          (31:2:002999)


 rec  pg cd   param code  control   param value
   2    3         00h       74h     0x0          (31:3:002999)
   2    3         01h       74h     0x0          (31:3:002999)
   2    3         02h       74h     0x0          (31:3:002999)
   2    3         03h       74h     0x0          (31:3:002999)
   2    3         04h       74h     0x0          (31:3:002999)
   2    3         05h       74h     0x190        (31:3:002999)
   2    3         06h       74h     0x0          (31:3:002999)


 rec  pg cd   param code  control   param value
   2    6         00h       74h     0x0          (31:6:002999)

<<<NOTE:  This output has been truncated for inclusion on this
man page.>>>
.fi
.ft
.SH FILES
.TP 20
\fB/var/opt/SUNWsamfs/sef/sefdata\fR
The default system error facility log file for Sun StorEdge \%SAM-FS and 
Sun \%SAM-QFS file systems.
.SH SEE ALSO
\fBmcf\fR(5),
\fBsefdata\fR(5),
\fBsefsysevent\fR(5).


src/utility/sefreport/sefreport.c
/*
 * sefreport.c
 */

/*
 *    SAM-QFS_notice_begin
 *
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
 * or https://illumos.org/license/CDDL.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at pkg/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 *
 *    SAM-QFS_notice_end
 */

#pragma ident "$Revision: 1.23 $"

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <time.h>
#include <string.h>
#include <strings.h>
#include <sys/int_types.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#define DEC_INIT
#include "sam/types.h"
#include "sam/lib.h"
#include "aml/sefstructs.h"
#include "aml/sefvals.h"
#include "aml/external_data.h"


#define	WRITE_ERROR_LOG_PG "Write Errors Counter"
#define	READ_ERROR_LOG_PG  "Read Errors Counter"
#define	NON_MEDIA_LOG_PG  "Non-Medium Error"

/* Write Error Counters log page 3. */
char *write_error_log_pg [] = {
	/* 0 */ "Errors corrected without substantial delay",
	/* 1 */ "Errors corrected with possible delays",
	/* 2 */ "Total rewrites",
	/* 3 */ "Total errors corrected",
	/* 4 */ "Total times correction algorithm processed",
	/* 5 */ "Total bytes processed",
	/* 6 */ "Total uncorrected errors"
};

/* Read Error Counters log page 3. */
char *read_error_log_pg [] = {
	/* 0 */ "Errors corrected without substantial delay",
	/* 1 */ "Errors corrected with possible delays",
	/* 2 */ "Total rewrites",
	/* 3 */ "Total errors corrected",
	/* 4 */ "Total times correction algorithm processed",
	/* 5 */ "Total bytes processed",
	/* 6 */ "Total uncorrected errors"
};

/* Non-Media Error log page 6. */
char *non_media_log_pg [] = {
	/* 0 */ "Non-medium error count"
};

int
main(
	int	argc,
	char *argv[]
)
{
	extern int	optind;
	void		usage();
	int			checkheader(struct sef_hdr *, int);
	int			checkrecord(char *, int, int);

	uchar_t		paramlen;
	char		*seffile = NULL;
	char		*timestr, *buf, *pageptr, paramstr[7];
	int			readcount, seffd, c, rec;
	int			i, bytesprinted, verbose = 0, devinfo = 0;
	int			desc = 0;
	uint16_t	pagelen, equip;
	uint16_t	paramcode;
	struct sef_hdr	header;
	struct sef_pg_hdr *pagehdr;
	int			field;
	struct stat	statbuf;
	char		**log_pg;
	int			paramcodemax;
	char		*pagename;
	char		*seffilep = NULL;


	while ((c = getopt(argc, argv, "d:vt")) != EOF) {
		switch (c) {
			case 'v':
				verbose = 1;
				break;
			case 'd':
				devinfo = 1;
				seffilep = optarg;
				break;
			case 't':
				desc = 1;
				break;
			case '?':
			default:
				usage(argv[0]);
				exit(1);
		} /* end switch */
	} /* end while */

	if (verbose) {
		desc = 0;
	}

	if (seffilep == NULL) {
		/* No file name */
		usage(argv[0]);
		exit(1);
	}

	seffile = strdup(seffilep);

	if (optind != argc) {
		/* extra stuff at end of command */
		usage(argv[0]);
		exit(1);
	}

	if ((seffd = open(seffile, O_RDONLY)) < 0) {
		printf("Error opening <%s>, errno <%d>.\n",
		    seffile, errno);
		exit(1);
	}

	if (fstat(seffd, &statbuf) < 0) {
		printf("Error determing filesize <%s>, errno <%d>.\n",
		    seffile, errno);
		exit(1);
	}

	if (statbuf.st_size <= 0) {
		/* 0 byte length sefdata file. */
		exit(0);
	}

	/*
	 * Read first sef record header.  This is struct sef_hdr.
	 */
	readcount = read(seffd, &header, sizeof (struct sef_hdr));
	if (checkheader(&header, readcount) < 0) {
		exit(1);
	}
	rec = 1;

	while (readcount != 0) {

		timestr = strdup(ctime(&header.sef_timestamp));
		/* get rid of annoying trailing \n that ctime puts on */
		timestr[strlen(timestr)-1] = '\0';
		printf("\nRecord no. %d\n", rec);
		printf("%s  %s %s %s VSN %s\n", timestr, header.sef_vendor_id,
		    header.sef_product_id, header.sef_revision, header.sef_vsn);
		BE16toH(&header.sef_eq, &equip);
		if (devinfo) {
			printf("   Eq no. %d   Dev name %s\n",
			    equip, header.sef_devname);
		}
		printf("\n");
		/* end of sef record header contents */

		/*
		 * the sef_size field in the record header is
		 * the total size of all pages in this record
		 */
		if ((buf = (char *)malloc(header.sef_size)) == NULL) {
			printf("ERROR:  Cannot malloc; errno %d.\n", errno);
			exit(1);
		}

		readcount = read(seffd, buf, header.sef_size);
		if (checkrecord(buf, readcount, header.sef_size) < 0) {
			exit(1);
		}

		pagehdr = (struct sef_pg_hdr *)buf;
		while (readcount > 0) {
			BE16toH(&pagehdr->page_len, &pagelen);


			/* setup log page decode text. */
			log_pg = NULL;
			paramcodemax = -1;
			pagename = NULL;
			if (pagehdr->page_code == SEF_WR_ERR_LOG_PG) {
				pagename = WRITE_ERROR_LOG_PG;
				log_pg = write_error_log_pg;
				paramcodemax = 6;
			} else if (pagehdr->page_code == SEF_RD_ERR_LOG_PG) {
				pagename = READ_ERROR_LOG_PG;
				log_pg = read_error_log_pg;
				paramcodemax = 6;
			} else if (pagehdr->page_code == SEF_NON_MEDIA_LOG_PG) {
				pagename = NON_MEDIA_LOG_PG;
				log_pg = non_media_log_pg;
				paramcodemax = 0;
			}

			if (!verbose) {
				printf("   PAGE CODE %x", pagehdr->page_code);
				if (pagename && desc) {
					printf(" %s", pagename);
				}
				printf("\n");
			} else {
				printf(" rec  pg cd");
			}

			/* print contents of page */
			printf("   param code  control   param value");
			if (desc) {
				printf("  description");
			}
			printf("\n");

			pageptr = (char *)pagehdr + sizeof (struct sef_pg_hdr);
			bytesprinted = 0;
			while (bytesprinted < pagelen) {
				if (verbose) {
					/*
					 * print record number and page code
					 * on each line
					 */
					printf("%4d  %3x  ",
					    rec, pagehdr->page_code);
				}
				/*
				 * print parameter code -- 2 bytes long,
				 * ranging in value from 01h to 8000h
				 */
				paramcode = (pageptr[0] << 8) | pageptr[1];
				sprintf(paramstr, "%02xh", paramcode);
				printf("   %7s   ", paramstr);
				pageptr = (char *)pageptr + 2;
				bytesprinted += 2;

				/*
				 * print control byte -- one byte
				 * print as upper 4 bits followed by
				 * lower 4 bits
				 */
				printf("    %x%xh  ", *(uchar_t *)pageptr >> 4,
				    *(uchar_t *)pageptr & 0xF);
				pageptr = (char *)pageptr + 1;
				bytesprinted += 1;

				/*
				 * print paramter value.  length varies
				 */
				paramlen = *(uchar_t *)pageptr;
				/*
				 * Advance pageptr past parameter length byte
				 */
				pageptr = (char *)pageptr + 1;
				bytesprinted += 1;

				/*
				 * Avoid odd byte memory alignment. Display
				 * paramdata without leading zeros. If
				 * paramdata is zero, then display one zero.
				 */
				printf("   0x");
				field = 0;
				for (i = 0; i < paramlen; i++) {
					if (field == 0 && pageptr [i] != 0) {
						/* First byte of number. */
						printf("%x",
						    (uchar_t)pageptr [i]);
						field = ((pageptr [i] >> 4) !=
						    0 ? 2 : 1);
					} else if (field != 0) {
						/*
						 * Pad middle of number with
						 * zeros.
						 */
						printf("%02x",
						    (uchar_t)pageptr [i]);
						field += 2;
					}
				}
				if (field == 0) {
					/* Zero number. */
					printf("0");
					field = 1;
				}

				/* fill out param value field width. */
				for (i = field; i < 11; i++) {
					printf(" ");
				}

				if (verbose) {
					printf("(%d:%x:%s)",
					    equip, pagehdr->page_code,
					    header.sef_vsn);
				} else if (log_pg && desc &&
				    (paramcode >= 0 && paramcode <=
				    paramcodemax)) {
					printf("%s", log_pg [paramcode]);
				}

				pageptr = (char *)pageptr + (int)paramlen;
				bytesprinted += (int)paramlen;

				printf("\n");
			} /* end while */  /* one log sense page printed */

			printf("\n\n");
			/* advance pagehdr pointer */
			pagehdr = (struct sef_pg_hdr *)((char *)pagehdr +
			    sizeof (struct sef_pg_hdr) + pagelen);
			readcount -= (sizeof (struct sef_pg_hdr) + pagelen);
		} /* end while */


		/*
		 * Free the malloc'd buffer for this sef record (these
		 * log sense pages).
		 */
		free(buf);

		/*
		 * Read next sef record header.  This is struct sef_hdr.
		 */
		readcount = read(seffd, &header, sizeof (struct sef_hdr));
		if (checkheader(&header, readcount) < 0) {
			exit(1);
		}
		rec++;

	} /* end while */

	return (0);

} /* end main */

int
checkheader(
	struct sef_hdr *hdr,
	int	bytecount
)
{
	if (bytecount < 0) {
		printf("ERROR:  Error reading sef file; errno %d.\n", errno);
		return (-1);
	}

	if ((bytecount > 0) && (bytecount != sizeof (struct sef_hdr))) {
		printf("ERROR:  Read %d bytes, expected %d.\n", bytecount,
		    sizeof (struct sef_hdr));
		return (-1);
	}

	if (hdr->sef_magic != SEFMAGIC) {
		printf("ERROR:  Bad magic.\n");
		return (-1);
	}

	if (hdr->sef_version != SEFVERSION) {
		printf("ERROR:  Bad version.  Got %d, expected %d.\n",
		    hdr->sef_version, SEFVERSION);
		return (-1);
	}

	return (0);
} /* end checkheader() */

void
usage(char *s)
{
	printf("USAGE:  %s [-v|-t] -d file\n", s);
}

int
checkrecord(
	char *buffer,
	int bytesread,
	int sizeexpected
)
{
	struct sef_pg_hdr *phdr;

	if (bytesread == 0) {
		printf("ERROR:  Unexpected end of file reached.\n");
		return (-1);
	}

	if (bytesread != sizeexpected) {
		printf("ERROR:  Read %d bytes, expected %d.\n",
		    bytesread, sizeexpected);
		return (-1);
	}

	phdr = (struct sef_pg_hdr *)buffer;
	if ((phdr->page_code < 0) || (phdr->page_len <= 0)) {
		printf("ERROR:  Log sense page header contents "
		    "make no sense.\n");
		return (-1);
	}

	return (0);
}
